src/invert/laplace/impls/petsc3damg/petsc3damg.hxx

```diff
@@ -0,0 +1,228 @@
+/**************************************************************************
+ * 3D Laplace inversion using PETSc solvers with algebraic multigrid
+ *  preconditioning
+ *
+ * Equation solved is: \f$d\nabla^2_\perp x + (1/c1)\nabla_perp c2\cdot\nabla_\perp x + ex\nabla_x x + ez\nabla_z x + a x = b\f$
+ *
+ **************************************************************************
+ * Copyright 2019 J. Buchanan, J. Omotani, C. MacMackin
+ *
+ * Contact: Ben Dudson, bd512@york.ac.uk
+ *
+ * This file is part of BOUT++.
+ *
+ * BOUT++ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+class LaplacePetsc3dAmg;
+
+#ifndef __PETSC_LAPLACE_3DAMG_H__
+#define __PETSC_LAPLACE_3DAMG_H__
+
+#ifdef BOUT_HAS_PETSC
+
+#include <globals.hxx>
+#include <output.hxx>
+#include <options.hxx>
+#include <invert_laplace.hxx>
+#include <boutexception.hxx>
+#include <bout/operatorstencil.hxx>
+#include <bout/petsclib.hxx>
+#include <bout/petsc_interface.hxx>
+#include <petscksp.h>
+
+class LaplacePetsc3dAmg;
+
+namespace {
+RegisterLaplace<LaplacePetsc3dAmg> registerlaplacepetsc3damg(LAPLACE_PETSC3DAMG);
+}
+
+class LaplacePetsc3dAmg : public Laplacian {
+public:
+  LaplacePetsc3dAmg(Options *opt = nullptr, const CELL_LOC loc = CELL_CENTRE, Mesh *mesh_in = nullptr);
+  virtual ~LaplacePetsc3dAmg() override;
```

Comment:
Probably a copy-paste typo, but `virtual` isn't needed if `override` is present

A good rule is [every virtual method should have exactly one of `virtual`, `override` or `final`](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final)